### PR TITLE
Adds train_batch_size, eval_batch_size, and n_gpu to to_sanitized_dict output for logging.

### LINKED ISSUE #5330: Better hyperparameter tensorboard logging in Trainer.
# 🚀 Feature request

`Trainer` should write `train_batch_size` and `eval_batch_size` to tensorboard. Currently only `per_device_` batch sizes are logged as hyperparamters, which means unless you know how many gpus you trained on (which is also not logged), you can't know the actual batch sizes used in training.

## Your contribution
 
Submitted a PR: [Adds train_batch_size, eval_batch_size, and n_gpu to to_sanitized_dict output for logging.](https://github.com/huggingface/transformers/pull/5331#issue-440857548)

### PR DESCRIPTION
**Pull request by jaymody**
Closes #5330 

``` 
>>> from transformers import TrainingArguments
>>> args = TrainingArguments("dir")
>>> args.to_sanitized_dict()
{
    "output_dir": "dir",
    "overwrite_output_dir": False,
    "do_train": False,
    ...
    "train_batch_size": 8,
    "eval_batch_size": 8,
    "n_gpu": 0,
}
```

### CODE REVIEW

**Comment by jaymody** [comment]
It seems that this is also true for the wandb config logging, but instead it uses `vars(args)` instead of `args.to_sanitized_dict()`.  Is there any reason the following line is not using`to_sanitized_dict()`?

https://github.com/huggingface/transformers/blob/1af58c07064d8f4580909527a8f18de226b226ee/src/transformers/trainer.py#L330

Here's a diff between my new implementation of `arg.to_sanitized_dict()` and `vars(args)`:

```
>>> from  transformers import TrainingArguments
>>> args = TrainingArguments("some_dir")
>>> set(vars(args).items()) - set(args.to_sanitized_dict().items())
{
    ("__cached__setup_devices", (device(type="cpu"), 0)),
    ("tpu_num_cores", None),
    ("per_gpu_train_batch_size", None),
    ("per_gpu_eval_batch_size", None),
    ("save_total_limit", None),
}
>>> set(args.to_sanitized_dict().items()) - set(vars(args).items())
{
    ("tpu_num_cores", "None"),
    ("per_gpu_train_batch_size", "None"),
    ("train_batch_size", 8),
    ("n_gpu", 0),
    ("eval_batch_size", 8),
    ("per_gpu_eval_batch_size", "None"),
    ("save_total_limit", "None"),
}
```

The main difference is that in `to_sanitized_dict()` the `__cached_setup_devices` property is not saved, and instead of `None`, the value is stringified to `"None"` (and of course, with my new changes `train_batch_size`, `eval_batch_size`, and `n_gpu` are also recoreded).

Gonna add another commit to make the wandb config logging use `to_sanitized_dict`.

File: src/transformers/training_args.py
Comment (JetRunner) [inline src/transformers/training_args.py:219]: I think some attention needs to be paid here. When using distributed training, `self.n_gpu` would always be `1`, which causes ambiguity. Is it expected behavior?

Comment (jaymody) [reply src/transformers/training_args.py:219]: Hm, that's a good point. Maybe it's best to just leave out `n_gpu`. 

The purpose of this feature was to have all the necessary hparams to reproduce a result. I've been training models between a system with one gpu, and another one with multiple gpus, which meant unless I knew which system a model was trained on, I wouldn't know what the true batch size it was trained with.

We don't really need `n_gpu` to reproduce the result so I'll just remove it from the dict.

**Comment by jaymody** [comment]
Hi @JetRunner , was hoping I can get an update on this PR. The test fail seems to be unrelated to my committed code.

**Comment by JetRunner** [comment]
@jaymody OK please wait for another approval

**Review by LysandreJik — APPROVED** [review]
Sure, this LGTM, cool for reproducibility. Wdyt @sgugger ?

**Review by sgugger — APPROVED** [review]
LGTM too!

**Pull request merged**
